src/osdc/Objecter.cc

```diff
@@ -2390,6 +2390,16 @@ void Objecter::_op_submit(Op *op, shunique_lock<ceph::shared_mutex>& sul, ceph_t
   sl.unlock();
   put_session(s);
 
+  if ((cct->_conf->objecter_homeless_timeout > 0) && s->is_homeless()) { 
+    timer.add_event(ceph::make_timespan(cct->_conf->objecter_homeless_timeout), [this, tid, s]() {
+      if (s->is_homeless()) {
```

Comment:
what guarantee do we have on the lifetime of `OSDSession *s`? this doesn't look safe without some kind of reference counting

Reply:
I think that the safest way here is just fetching s again using this loop (it came from the _op_cancel): 
  for (auto siter = osd_sessions.begin();siter != osd_sessions.end(); ++siter) {
    OSDSession *s = siter->second;
    if (s->ops.find(tid) != s->ops.end()) {
      if ((cct->_conf->objecter_homeless_timeout > 0) && s->is_homeless()) { 
    timer.add_event(ceph::make_timespan(cct->_conf->objecter_homeless_timeout), [this, tid, s]() {
      if (s->is_homeless()) {
...

@cbodley what do you think?

Reply:
@cbodley Can you take a look please?

Reply:
I changed the location for the the cancellation and added cancel the timeout on session changed status.